managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerOfflineBacklog.java

```diff
@@ -102,369 +89,28 @@ private long getNumberOfEntries(Range<Position> range,
-                                            .setSize(lh.getLength()).setTimestamp(System.currentTimeMillis()).build();
-                                    ledgers.put(id, info);
-                                    mlMetaCounter.countDown();
-                                } else if (Errors.isNoSuchLedgerExistsException(rc)) {
-                                    log.warn("[{}] Ledger not found: {}", managedLedgerName, ledgers.lastKey());
-                                    ledgers.remove(ledgers.lastKey());
-                                    mlMetaCounter.countDown();
-                                } else {
-                                    log.error("[{}] Failed to open ledger {}: {}", managedLedgerName, id,
-                                            BKException.getMessage(rc));
-                                    mlMetaCounter.countDown();
-                                }
-                            };
-
-                            if (log.isDebugEnabled()) {
-                                log.debug("[{}] Opening ledger {}", managedLedgerName, id);
-                            }
-
-                            factory.getBookKeeper()
-                                    .thenAccept(bk -> {
-                                        bk.asyncOpenLedgerNoRecovery(id, digestType, password, opencb, null);
-                                    }).exceptionally(ex -> {
-                                        log.warn("[{}] Failed to open ledger {}: {}", managedLedgerName, id, ex);
-                                        opencb.openComplete(-1, null, null);
-                                        mlMetaCounter.countDown();
-                                        return null;
-                                    });
-                        } else {
-                            log.warn("[{}] Ledger list empty", managedLedgerName);
-                            mlMetaCounter.countDown();
-                        }
-                    }
-
-                    @Override
-                    public void operationFailed(ManagedLedgerException.MetaStoreException e) {
-                        log.warn("[{}] Unable to obtain managed ledger metadata - {}", managedLedgerName, e);
-                        mlMetaCounter.countDown();
-                    }
-                });
-
-        if (accurate) {
-            // block until however long it takes for operation to complete
-            mlMetaCounter.await();
+        if (factory instanceof ManagedLedgerFactoryImpl) {
+            List<Object> ctx = new ArrayList<>();
+            ctx.add(digestType);
+            ctx.add(password);
+            factory.estimateUnloadedTopicBacklog(offlineTopicStats, topicName, accurate, ctx);
         } else {
-            mlMetaCounter.await(META_READ_TIMEOUT_SECONDS, TimeUnit.SECONDS);
-
+            Object ctx = null;
+            factory.estimateUnloadedTopicBacklog(offlineTopicStats, topicName, accurate, ctx);
```

Comment:
This `estimateUnloadedTopicBacklog` API is not intuitive.
1. Why it needs an `accurate` parameter? (actually it's always false) 
2. If the factory is not a `ManagedLedgerFactoryImpl`, the `ctx` field is always null.

<img width="1051" alt="image" src="https://github.com/user-attachments/assets/1e17b59c-e54e-4b84-88db-391492cdcb8a">

It's not easy for the author of a 3rd party ML to know how to implement this method correctly without enough API docs. The API design is bad. It should be something like:

```java
   PersistentOfflineTopicStats estimateUnloadedTopicBacklog(TopicName topicName);
```

BTW, IMHO, this API should be moved to an internal method of `ManagedLedgerFactoryImpl`, but it might need another proposal to discuss about the API design.

Reply:
> BTW, IMHO, this API should be moved to an internal method of `ManagedLedgerFactoryImpl`, but it might need another proposal to discuss about the API design.

Yes, this is a mess that should be untangled. The usage is here: https://github.com/apache/pulsar/blob/faed541bc64ed112eb3c61abd9cdda10ecaeed1e/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java#L3144-L3184

Reply:
Since this PR's motivation is mainly to decouple the ML interface from the Impls, this comment can be improved in a future PR